### DISCUSSION TITLE
PSR SimpleCache connection error

### DISCUSSION THREAD

**Discussion by tricarte** [cachewerk/relay#129] [Support] [answered]
This is a sample code connecting to Redis using relay with its PSR16 cache implementation.

```php
<?php

require_once __DIR__ . '/vendor/autoload.php';

use Relay\Relay;
use CacheWerk\Relay\Psr\SimpleCache\RelayCache;

$relay = new Relay;
$relay->connect('127.0.0.1', 6379);

echo $relay->get('some:key') . "\n"; // This line works

$cache = new RelayCache($relay);
echo $cache->get('some:key') . "\n"; // This gives 'not connected' error
```

This is the full error:

```
PHP Fatal error:  Uncaught Relay\Exception: Not connected (RELAY_ERR_IO; relay.c:3513) in /home/user/repos/relay-redis/vendor/cachewerk/relay/src/Psr/SimpleCache/RelayCache.php:43
Stack trace:
#0 /home/user/repos/relay-redis/vendor/cachewerk/relay/src/Psr/SimpleCache/RelayCache.php(43): Relay\Relay->get()
#1 /home/user/repos/relay-redis/index.php(14): CacheWerk\Relay\Psr\SimpleCache\RelayCache->get()
#2 {main}

Next CacheWerk\Relay\Psr\SimpleCache\CacheException: Not connected (RELAY_ERR_IO; relay.c:3513) in /home/user/repos/relay-redis/vendor/cachewerk/relay/src/Psr/SimpleCache/RelayCache.php:47
Stack trace:
#0 /home/user/repos/relay-redis/index.php(14): CacheWerk\Relay\Psr\SimpleCache\RelayCache->get()
#1 {main}
  thrown in /home/user/repos/relay-redis/vendor/cachewerk/relay/src/Psr/SimpleCache/RelayCache.php on line 47
```

So, How am I supposed to construct the RelayCache instance?

**Comment by tillkruss** [comment]
@michael-grunder Can you elaborate on `RELAY_ERR_IO`?

**Reply by tricarte** [reply]
Sorry but that's the entire error message. I don't know any way to further dig into.

I tried to integrate Relay extension into a web framework that uses `symfony/cache` and `illuminate/redis` and they all worked correctly there. Then I realized that Relay has its own PSR-16 implementation but it didn't work in that framework. Only then I tried this simple code but I couldn't run it with success.

So, as I understand, there is nothing wrong with the code?

**Reply by tillkruss** [reply]
@tricarte: I meant @michael-grunder can tell us exactly what `Not connected (RELAY_ERR_IO)` means to diagnose this.

**Reply by tricarte** [reply]
@tillkruss Oh, one of those thousand ways to feel like stupid. Sorry!

**Reply by tillkruss** [reply, marked as answer]
Hehe, no worries. We've fixed the issue: https://github.com/cachewerk/relay/pull/131

**Comment by michael-grunder** [comment]
That error generally happens when we try to read data from Redis when not connected at all.  For example this would cause it:

```php
$relay = new \Relay\Relay;
$relay->get('foo');
```

So my guess is that the adapter isn't doing any sort of connection before trying to use it.  I'll see if I can replicate the behavior locally.

**Reply by michael-grunder** [reply]
This is fixed in #131

**Reply by tricarte** [reply]
Thanks for the fix and explanation.
